File: backend/app/services/engines.py
"""描画エンジンのレジストリ。

エンジンごとの属性（ゲート・PDF必須・無料枠カウント等）をENGINE_SPECSの1行に集約し、
判定に使う集合はすべてそこから導出する。エンジンの追加・変更はこの表の1行で完結させ、
判定ごとに列挙を書き足さない。
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Callable, FrozenSet


@dataclass(frozen=True)
class EngineSpec:
    name: str
    # LLMがHTML/CSS/JSONを生成する。falseはAIを介さず変換結果をそのまま描画結果にするエンジン。
    uses_ai: bool
    # 無料枠を超えるAPI利用が発生するため、ログイン済みユーザーにのみ提供する。
    gated: bool
    # 変換対象、またはhybridの統合入力としてPDFそのものが要る。
    requires_pdf: bool
    # gemini_freeと同じ無料枠モデルを使うため、全ユーザー共有の日次カウンタの対象になる。
    counts_free_quota: bool


ENGINE_SPECS: tuple[EngineSpec, ...] = (
    EngineSpec("gemini_free", uses_ai=True, gated=False, requires_pdf=False, counts_free_quota=True),
    EngineSpec("gemini", uses_ai=True, gated=True, requires_pdf=False, counts_free_quota=False),
    EngineSpec("claude", uses_ai=True, gated=True, requires_pdf=False, counts_free_quota=False),
    EngineSpec("openai", uses_ai=True, gated=True, requires_pdf=False, counts_free_quota=False),
    EngineSpec("hybrid", uses_ai=True, gated=False, requires_pdf=True, counts_free_quota=True),
    EngineSpec("docling", uses_ai=False, gated=False, requires_pdf=True, counts_free_quota=False),
    EngineSpec("pdf2htmlex", uses_ai=False, gated=False, requires_pdf=True, counts_free_quota=False),
    EngineSpec("pymupdf", uses_ai=False, gated=False, requires_pdf=True, counts_free_quota=False),
)


def _names_where(predicate: Callable[[EngineSpec], bool]) -> FrozenSet[str]:
    return frozenset(spec.name for spec in ENGINE_SPECS if predicate(spec))


# 未知のengine値をPDF読み込み等の重い処理より前に弾くための許可リスト。
ALL_ENGINES: FrozenSet[str] = _names_where(lambda spec: True)
AI_ENGINES: FrozenSet[str] = _names_where(lambda spec: spec.uses_ai)
CONVERTER_ENGINES: FrozenSet[str] = _names_where(lambda spec: not spec.uses_ai)
GATED_ENGINES: FrozenSet[str] = _names_where(lambda spec: spec.gated)
PDF_REQUIRED_ENGINES: FrozenSet[str] = _names_where(lambda spec: spec.requires_pdf)
GEMINI_FREE_QUOTA_ENGINES: FrozenSet[str] = _names_where(lambda spec: spec.counts_free_quota)

File: frontend/src/lib/engines.ts
import { Bot, Brain, FileCode2, FileText, Gem, Layers, Rows3, Sparkles } from 'lucide-react'
import type { ComponentType } from 'react'

// backendのapp/services/engines.pyのENGINE_SPECSと同じ値。
export type RenderEngineId =
  | 'gemini_free'
  | 'gemini'
  | 'claude'
  | 'openai'
  | 'hybrid'
  | 'docling'
  | 'pdf2htmlex'
  | 'pymupdf'

export type EngineDefinition = {
  id: RenderEngineId
  label: string
  description: string
  icon: ComponentType<{ className?: string }>
  // 未ログインでは使えない標準プランの生成AI。判定をバックエンドに一本化するため、選択自体は
  // 許可し、描画を押した時点の403をメッセージとして表示する。
  gated: boolean
  // LLMがHTML/CSS/JSONを生成する。生成は20〜60秒以上かかることがあり、API Gatewayの29秒固定
  // タイムアウトを受けるため、このエンジンだけ非同期ジョブ＋ポーリングで描画する。
  usesAi: boolean
  // gemini_freeと同じ無料枠モデルを使うため、共有カウンタの表示対象になる。
  countsFreeQuota: boolean
}

// 描画エンジンの一覧。表示情報（ラベル・説明・アイコン）と挙動の分岐条件を1行にまとめ、
// エンジンの追加・変更をこの表だけで完結させる。
export const ENGINES: readonly EngineDefinition[] = [
  {
    id: 'hybrid',
    label: '精密復元',
    description: 'PyMuPDF・Docling・Gemini（無料枠）（数十秒〜数分）',
    icon: Layers,
    gated: false,
    usesAi: true,
    countsFreeQuota: true,
  },
  {
    id: 'gemini_free',
    label: 'Gemini API（無料枠）',
    description: 'PDFを直接読み取り、無料枠モデルで整形します',
    icon: Sparkles,
    gated: false,
    usesAi: true,
    countsFreeQuota: true,
  },
  {
    id: 'gemini',
    label: 'Gemini API',
    description: '標準プランのモデルでより高精度に整形します',
    icon: Gem,
    gated: true,
    usesAi: true,
    countsFreeQuota: false,
  },
  {
    id: 'claude',
    label: 'Claude API',
    description: 'Anthropic Claudeで高精度に整形します',
    icon: Bot,
    gated: true,
    usesAi: true,
    countsFreeQuota: false,
  },
  {
    id: 'openai',
    label: 'OpenAI API',
    description: 'OpenAIのモデルで高精度に整形します',
    icon: Brain,
    gated: true,
    usesAi: true,
    countsFreeQuota: false,
  },
  {
    id: 'docling',
    label: 'Docling',
    description: 'PDFのテキストと構造をHTML化します（AIなし）',
    icon: FileText,
    gated: false,
    usesAi: false,
    countsFreeQuota: false,
  },
  {
    id: 'pdf2htmlex',
    label: 'pdf2htmlEX',
    description: 'PDFの見た目をそのままHTML化します（AIなし）',
    icon: FileCode2,
    gated: false,
    usesAi: false,
    countsFreeQuota: false,
  },
  {
    id: 'pymupdf',
    label: 'PyMuPDF',
    description: 'PDFのレイアウトを座標付きで再現します（AIなし）',
    icon: Rows3,
    gated: false,
    usesAi: false,
    countsFreeQuota: false,
  },
]

function idsWhere(predicate: (engine: EngineDefinition) => boolean): ReadonlySet<RenderEngineId> {
  return new Set(ENGINES.filter(predicate).map((engine) => engine.id))
}

export const AI_ENGINES = idsWhere((engine) => engine.usesAi)
export const GEMINI_FREE_QUOTA_ENGINES = idsWhere((engine) => engine.countsFreeQuota)
